src/corelibs/U2Formats/src/FastaFormat.cpp

```diff
@@ -70,180 +54,184 @@ FastaFormat::FastaFormat(QObject *p)
+            if (currentSequenceLength > 0) {
+                minLen = minLen == -1 ? currentSequenceLength : qMin(minLen, currentSequenceLength);
+                maxLen = maxLen == -1 ? currentSequenceLength : qMax(maxLen, currentSequenceLength);
             }
-            len = 0;
+            currentSequenceLength = 0;
         } else {
-            len += line.length();
+            currentSequenceLength += line.length();
             if (!hasGaps && line.contains(U2Msa::GAP_CHAR)) {
                 hasGaps = true;
             }
         }
-    } while (!input.atEnd());
+    } while (!textStream.atEnd());
 
-    QVariantMap res;
-    res[RawDataCheckResult_Sequence] = true;
+    QVariantMap hints;
+    hints[RawDataCheckResult_Sequence] = true;
     if (hasGaps) {
-        res[RawDataCheckResult_SequenceWithGaps] = true;
+        hints[RawDataCheckResult_SequenceWithGaps] = true;
     }
     if (nSequences > 1) {
-        res[RawDataCheckResult_MultipleSequences] = true;
+        hints[RawDataCheckResult_MultipleSequences] = true;
     }
     if (minLen > 0) {
-        res[RawDataCheckResult_MaxSequenceSize] = maxLen;
-        res[RawDataCheckResult_MinSequenceSize] = minLen;
+        hints[RawDataCheckResult_MaxSequenceSize] = maxLen;
+        hints[RawDataCheckResult_MinSequenceSize] = minLen;
     }
-    return res;
+    return hints;
 }
 
-FormatCheckResult FastaFormat::checkRawTextData(const QByteArray &rawData, const GUrl &) const {
-    const char *data = rawData.constData();
-    int size = rawData.size();
-
-    int n = TextUtils::skip(TextUtils::WHITES, data, size);
-    int newSize = size - n;
-    const char *newData = data + n;
-    if (newSize <= 0 || (newData[0] != FASTA_HEADER_START_SYMBOL && newData[0] != FASTA_COMMENT_START_SYMBOL)) {
+FormatCheckResult FastaFormat::checkRawTextData(const QString &dataPrefix, const GUrl &) const {
+    QString data = TextUtils::skip(TextUtils::WHITES, dataPrefix);
+    FormatDetectionScore score;
+    if (data.startsWith(FASTA_HEADER_START_SYMBOL)) {
+        // A perfect match: start of the FASTA file.
+        score = FormatDetection_Matched;
+    } else if (data.startsWith(FASTA_COMMENT_START_SYMBOL) && data.contains(data.startsWith(QString("\n") + FASTA_HEADER_START_SYMBOL))) {
+        // A good match: looks like a FASTA comment before a FASTA header line.
+        score = FormatDetection_HighSimilarity;
+    } else {
         return FormatDetection_NotMatched;
     }
-    bool hasBinaryBlocks = TextUtils::contains(TextUtils::BINARY, data, size);
-    if (hasBinaryBlocks) {
-        return FormatDetection_NotMatched;
+    // Ok, format is matched -> add hints about sequence sizes.
+    FormatCheckResult result(score);
+    result.properties = buildFormatDetectionHints(data);
+    return result;
+}
+
+/** Returns true if the line is a valid comment line for FASTA format. */
+static bool isCommentLine(const QString &line) {
```

Comment:
I think, it'll be better to define all these `static` functions in the .h file

Reply:
Headers is a part of public API. Local methods should stay local and do not spoil public interface or generate extra dependencies.

To make it possible there are different patterns in C++, like PImpl (used across QT code) or local functions.
So local static functions is a good thing that keeps local stuff local - I think we should use them all the time where it is possible.